### ISSUE TITLE
[Crons] Time outs do not create issues

### ISSUE THREAD

**Issue by InterstellarStella** [getsentry/sentry#59431]
### Environment

SaaS (https://sentry.io/)

### Steps to Reproduce

For testing purposes:
1. Set up a cron monitor for a 3 minute interval, 5 minute grace period and 7 minute max runtime, and set it to notify your email
2. Use postman and [this endpoint](https://docs.sentry.io/api/crons/create-a-new-checkin/) to create an `in_progress` checkin
![image](https://github.com/getsentry/sentry/assets/81985895/4ce0d14a-6e01-4782-86fa-ae758499776d)

3. Wait until the In Progress status turns to Timed Out [in the Sentry Web UI](https://dragonfruit-dev.sentry.io/crons/1340-5y/?project=4504490632871936&statsPeriod=90d)

### Expected Result

After 7 minutes the checkin should be marked as Timed Out, create an issue, and notify your email.

### Actual Result

No issue is created due to the time out, and no email notification is sent. This has caused a customer to miss a multi-hour failure.

I also tested the same thing with missed and failed checkins, both of which seem to be creating new issues and sending an email. If we mark those issues as resolved, and test the same situations again, the issues will be marked as regressed, so this part works fine.

Links from my test org:
- [monitor](https://dragonfruit-dev.sentry.io/crons/1340-5y/edit/)
- [missed checkin issue](https://dragonfruit-dev.sentry.io/issues/4605859670/?project=4504490632871936&query=is%3Aunresolved&referrer=issue-stream&statsPeriod=90d&stream_index=0)
- [failed checkin issue](https://dragonfruit-dev.sentry.io/issues/4605897020/?project=4504490632871936&query=is%3Aunresolved&referrer=issue-stream&statsPeriod=90d&stream_index=1)

Links to the customer's org are provided in the shadow ticket.

### Product Area

Crons

┆Issue is synchronized with this [Jira Improvement](https://getsentry.atlassian.net/browse/FEEDBACK-1980) by [Unito](https://www.unito.io)

**Comment by rjo100** [comment]
Will be fixed by https://github.com/getsentry/sentry/pull/59445

**Comment by InterstellarStella** [comment]
It looks like this is happening again. I purposely created Timed Out check-ins [here](https://dragonfruit-dev.sentry.io/crons/test-monitor/?statsPeriod=90d) but no issue was created, which means I can't be notified about them either.

**Comment by wedamija** [comment]
Hi @InterstellarStella, sorry for the delay in response here.

I'm looking at the cron you linked to, and it seems like a missed checkin issue was created here: https://dragonfruit-dev.sentry.io/issues/4905276569/. This looks like it happened around the right day. If that's not what you expected, let me know and I can dig into this further

**Comment by InterstellarStella** [comment]
Hi @wedamija, I did this once more (and then disabled the monitor), and as you can see [here](https://dragonfruit-dev.sentry.io/crons/1340-5y/?project=4504490632871936&statsPeriod=90d), there is a time out but no issue associated to it. Could you take a look once again?

**Comment by gaprl** [comment]
Hey @InterstellarStella, I noticed the URL for the endpoint referenced in the issue is pointing to our legacy endpoints. Could you give it another try with our latest ingest endpoints? We have a guide on migrating [here](https://docs.sentry.io/product/crons/legacy-endpoint-migration/).

I just ran a test with the ingest-based endpoints and was able to see a timeout issue be created. Let me know if you see different results. Thank you.

**Comment by InterstellarStella** [comment]
@gaprl You're right, I was using the old endpoint. Used the new one but unfortunately [the issue remains](https://dragonfruit-dev.sentry.io/crons/1340-5y/?statsPeriod=90d). Tried it [in another org](https://dev-curumas.sentry.io/crons/helloflask/?project=5265761&statsPeriod=90d) too.

**Comment by wedamija** [comment]
@InterstellarStella Both of these monitors are muted, which is why they're not creating related issues. When I tested this in your test org it worked ok https://dragonfruit-dev.sentry.io/crons/dan-test-cron/?statsPeriod=1h

Could you confirm whether the customers monitor is muted?

**Comment by wedamija** [comment]
Ok, I understand the problem now.

https://github.com/getsentry/sentry/blob/2082a5c98afdbdcf14801ff35b79a7ca3829909c/src/sentry/monitors/logic/mark_failed.py#L175

Any monitors that were missed/failing before we switched over to issue platform, and are still in a failed state, won't successfully create an issue platform issue. I'll look into correcting this next week, but a stopgap is to make the monitor succeed once and enter a good state. I was able to get your monitor to produce a failed issue this way: https://dragonfruit-dev.sentry.io/issues/5048713934/?project=4504490632871936&query=is%3Aunresolved+issue.type%3Amonitor_check_in_timeout&referrer=issue-stream&statsPeriod=90d&stream_index=0

**Comment by gaprl** [comment]
Closing this as resolved. If you are experiencing a similar issue, please do as mentioned above: create a succesful check-in, and then any issues afterwards should be created properly.

**Comment by InterstellarStella** [comment]
@wedamija I created a new monitor [here](https://curumas-sponsored.sentry.io/crons/cronproject/will-it-create-an-issue/) and made sure it got a successful check in first. I then let it get a time out, which did not create an issue. The following 4 missed check ins correctly created an issue with 4 events. (note that I have now paused the monitor since I saw what I needed to see)
The question in, why aren't issues for timeouts being created? Should issues for timeouts be created in the first place?

**Comment by wedamija** [comment]
> @wedamija I created a new monitor [here](https://curumas-sponsored.sentry.io/crons/cronproject/will-it-create-an-issue/) and made sure it got a successful check in first. I then let it get a time out, which did not create an issue. The following 4 missed check ins correctly created an issue with 4 events. (note that I have now paused the monitor since I saw what I needed to see) The question is, why aren't issues for timeouts being created? Should issues for timeouts be created in the first place?

Hey @InterstellarStella. Looking at your config, your monitor is set up to timeout after 5 minutes, and to miss after 2 minutes. 

So when I look at your checkins I see
![Screenshot 2024-04-23 at 2 45 29 PM](https://github.com/getsentry/sentry/assets/6288560/44a90f33-76d3-4bf7-863e-1b784e3d59f6)

Looking in the database, some of the dates are a little different to what we display here, so that's something we should improve. But the order of operations is

So the order of operations is:
 - 12:38:26 AM PDT Receive an in progress 
 - 12:41:00 AM PDT Detected a miss
 - 12:43:00 AM PDT Detected a miss
 - 12:43:00 AM PDT In progress checkin times out

Since the misses happen sequentially before we actually detect the timeout, the issue that gets created is associated with the misses. When the issue finally times out, the issue is already created, and due to the fact that it started before the misses, we end up not including the timeout in the count of the created issue.

This is a little tricky to resolve, because I'm not totally clear if it's a bug. It might actually make more sense on our end to prevent timeouts that are larger than the interval, since typically you don't want a job that fires every minute to take longer than a minute (although maybe there are exceptions).

I'm not sure if this is the same issue our customer experienced, but this is at least the explanation for what you're seeing here.

**Comment by DaveRosenthalSentry** [comment]
Moving this from Bug to Improvement since the system is working as designed (albeit not ideally).
